src/modules/mavlink/streams/BATTERY_INFO.hpp

```diff
@@ -57,21 +58,43 @@ class MavlinkStreamBatteryInfo : public MavlinkStream
 	explicit MavlinkStreamBatteryInfo(Mavlink *mavlink) : MavlinkStream(mavlink) {}
 
 	uORB::SubscriptionMultiArray<battery_status_s, battery_status_s::MAX_INSTANCES> _battery_status_subs{ORB_ID::battery_status};
+	uORB::SubscriptionMultiArray<battery_info_s, battery_status_s::MAX_INSTANCES> _battery_info_subs{ORB_ID::battery_info};
+
+	uint8_t _serial_number_ids[battery_status_s::MAX_INSTANCES] {};
+	char _serial_numbers[battery_status_s::MAX_INSTANCES][32] {}; ///< keep track of serial numbers until all static battery info is split out
 
 	bool send() override
 	{
 		bool updated = false;
 
+		for (int i = 0; i < _battery_info_subs.size(); ++i) {
+			battery_info_s battery_info;
+
+			if (_battery_info_subs[i].update(&battery_info)) {
+				_serial_number_ids[i] = battery_info.id;
+				memcpy(_serial_numbers[i], battery_info.serial_number, sizeof(_serial_numbers[0]));
+			}
+		}
+
 		for (auto &battery_sub : _battery_status_subs) {
 			battery_status_s battery_status;
 
 			if (battery_sub.update(&battery_status)) {
-				if (battery_status.serial_number == 0) {
-					// Required to emit
-					continue;
+				mavlink_battery_info_t msg{};
+				bool battery_has_serial_number = false;
+
+				// load serial number from battery_info message until all static information fields were moved
+				for (int i = 0; i < battery_status_s::MAX_INSTANCES; ++i) {
+					if ((_serial_number_ids[i] != 0) && (_serial_number_ids[i] == battery_status.id)) {
+						snprintf(msg.serial_number, sizeof(msg.serial_number), "%s", _serial_numbers[i]);
+						battery_has_serial_number = true;
+					}
 				}
 
-				mavlink_battery_info_t msg{};
+				if (!battery_has_serial_number) {
+					// Only publish BATTERY_INFO if the battery has a serial number
+					continue;
+				}
 
 				msg.id = battery_status.id - 1;
 				msg.design_capacity = (float)(battery_status.capacity * 1000);
```

Comment:
1. The "invalid" values for these need to be checked (might not match earlier messages).
```
full_charge_capacity	float	Ah	invalid:NaN
cycle_count	uint16_t		invalid:UINT16_MAX
state_of_health	uint8_t	%	invalid:UINT8_MAX
```

Reply:
Did you have a "proper" code reviewer in mind? If not, I'd volunteer @julianoes :-)

Reply:
I'll find someone for review :pray: